### PR TITLE
Bug 1858948 - Add shopping NotAvailable product state

### PR DESCRIPTION
**Pull request by calumozilla**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1858948

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt [hunk 1]
```diff
@@ -5,8 +5,8 @@
...
 
 /**
- * Converts a string to an enum value, ignoring case. If the string does not match any of the
- * enum values, the default value is returned.
+ * Converts a string to an enum value, ignoring case and replacing spaces with underscores.
+ * If the string does not match any of the enum values, the default value is returned.
  */
 inline fun <reified T : Enum<T>> String.asEnumOrDefault(defaultValue: T? = null): T? =
-    enumValues<T>().firstOrNull { it.name.equals(this, ignoreCase = true) } ?: defaultValue
+    enumValues<T>().firstOrNull { it.name.equals(this.replace(" ", "_"), ignoreCase = true) } ?: defaultValue
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt:12]: Nice! Let's add tests for this

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt [hunk 2]
```diff
@@ -29,6 +29,10 @@ private fun ProductAnalysis.toProductReview(): ProductReviewState =
         }
     } else if (notEnoughReviews && !needsAnalysis) {
         ProductReviewState.Error.NotEnoughReviews
+    } else if (deletedProduct) {
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt:32]: Can this be true and `notEnoughReviews` is also true? If so, in that case, this'll map to `NotEnoughReviews`, should that happen or `NotAvailable`?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckState.kt [hunk 3]
```diff
@@ -159,6 +159,17 @@ sealed interface ReviewQualityCheckState : State {
...
                 }
             }
+
+            /**
+             * Denotes a product is not available.
+             *
+             * @property reported Whether or not current user has reported the product is back in stock.
+             * @property anotherUserReported Whether or not another user has reported the product is back in stock.
+             */
+            data class NotAvailable(
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckState.kt:169]: This should be a child of the Error sealed interface, similar to `NotEnoughReviews` and `UnsupportedProductTypeError`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt [hunk 4]
```diff
@@ -214,9 +219,4 @@ sealed interface ReviewQualityCheckAction : Action {
      * Triggered when the No analysis card is displayed to the user.
      */
     object NoAnalysisDisplayed : TelemetryAction
-
-    /**
-     * Triggered when the user reports a product is back in stock.
-     */
-    object ReportProductBackInStock : TelemetryAction
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt:221]: Let's use the existing action

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt [hunk 5]
```diff
@@ -216,3 +240,28 @@ enum class AnalysisStatusDto {
...
+
+    /**
+     * Product was not actually marked as deleted.
+     */
+    NOT_DELETED,
+
+    /**
+     * Any other status.
+     */
+    OTHER,
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:266]: Since this API has 3 states and we handle them specifically, we don't need to define this `Other` state IMO

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckState.kt (hunk 3 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckState.kt:169]: Also, Looking at the [designs](https://www.figma.com/file/VE5xayz3yOSB5qOATUU9Im/Mobile-Shopping-Experience-MVP?type=design&node-id=5384-23225&mode=design&t=l1duuWVGcjQZAxvB-0) we want to represent 3 different states, 

`ProductNotAvailable` -> When `deleted_product` is returned true by the analysis.  Shows the button also. 
`ThanksForReporting` -> When button is clicked on `ProductNotAvailable`
`ProductAlreadyReported` -> When `deleted_product_reported` is returned true by the analysis api.

With this, they become objects in the Error sealed interface, similar to other Error states and the UI is 1:1 mapped.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt (hunk 2 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt:32]: Based on this [comment](https://github.com/mozilla-mobile/firefox-android/pull/4756/files#r1422911249), we should also check for when `deletedProductReported` is true. The above question/clarification applies for this also.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStore.kt [hunk 6]
```diff
@@ -161,5 +161,17 @@ private fun mapStateForUpdateAction(
                 }
             }
         }
+
+        ReviewQualityCheckAction.ReportBackInStock -> {
+            state.mapIfOptedIn {
+                if (it.productReviewState is ProductReviewState.NotAvailable) {
+                    it.copy(
+                        productReviewState = it.productReviewState.copy(reported = true),
```
[the remark this answers is not in the document]
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStore.kt:169]: Check [this](https://github.com/mozilla-mobile/firefox-android/pull/4756#discussion_r1422912955) comment and the comment about representing 3 states.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt (hunk 1 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt:12]: Added a check in EnumMapperTest

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt (hunk 2 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt]: O good point, moved up the `else if (deletedProduct)` before checking `notEnoughReviews` and also added check for `deletedProductReported`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt (hunk 5 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt]: Sure, removed `OTHER` and set the default for `asEnumOrDefault` to `NOT_DELETED`, although I'm not sure if that's the best default between the 3.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt (hunk 4 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt]: Updated to use action `ReportProductBackInStock`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckState.kt (hunk 3 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckState.kt]: Updated the 3 states to be Error states

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStore.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStore.kt]: Reopening the bottom sheet now shows correct state

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt [hunk 7]
```diff
@@ -129,6 +88,95 @@ fun ProductAnalysisError(
     }
 }
 
+@Suppress("LongMethod")
```
Comment (calumozilla) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt:91]: had to suppress longMethod, maybe the ReviewQualityCheckInfoType could be set to Info to start?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt (hunk 5 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt]: We don't need a default since API returns one of the 3 strings/enum values. If it's something else, then the service will return null which is handled in the middleware.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt [hunk 8]
```diff
@@ -83,6 +83,13 @@ class ReviewQualityCheckNetworkMiddleware(
                     store.onReanalyze()
                 }
 
+                ReviewQualityCheckAction.ReportProductBackInStock -> {
+                    val status = reviewQualityCheckService.reportBackInStock()
+                    if (status == ReportBackInStockStatusDto.REPORT_CREATED) {
+                        store.updateProductReviewState(ProductReviewState.Error.ThanksForReporting)
+                    }
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]: This is fine for now but we should consider a future improvement here. 

When the user clicks the button, nothing really happens on the ui till the network request succeeds with the right response and then the UI transitions to thanks for reporting. Maybe we should add a loading state here while the request is inflight.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt [hunk 9]
```diff
@@ -195,6 +202,23 @@ class DefaultReviewQualityCheckService(
...
         }
     }
+
+    override suspend fun reportBackInStock(): ReportBackInStockStatusDto? = withContext(Dispatchers.Main) {
+        suspendCoroutine { continuation ->
+            browserStore.state.selectedTab?.let { tab ->
+                tab.engineState.engineSession?.reportBackInStock(
+                    url = tab.content.url,
+                    onResult = {
+                        continuation.resume(it.asEnumOrDefault(ReportBackInStockStatusDto.NOT_DELETED))
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:212]: Let's not set a default value in `asEnumOrDefault`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt [hunk 10]
```diff
@@ -232,3 +256,23 @@ data class AnalysisStatusProgressDto(
...
     val progress: Double,
 )
+
+/**
+ * Enum that represents the status returned from reporting a product is back in stock.
+ */
+enum class ReportBackInStockStatusDto {
+    /**
+     * Created reported.
+     */
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:266]: 😄 typo?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 8 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]: Or we should treat this as a success by default and update to `ThanksForReporting` immediately, and only revert back to the previous state in case of error.

Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]: It seems like that's already happening in the Store reducer, which now makes me wonder why do we need `store.updateProductReviewState(ProductReviewState.Error.ThanksForReporting)` ?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt (hunk 7 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt:91]: I see, let's keep this when logic in the above function in that case. 

We don't want to set a var and then overwrite in, it's better to be declarative and have one state.

I think Triple has fulfilled it's usecase here, let's create a data class with 5 properties that are required for the `InfoCard` composable. Then each error state maps to that object which is a param for `ReviewQualityCheckInfoCard`.

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt [hunk 11]
```diff
@@ -1282,6 +1282,55 @@ class ReviewQualityCheckStoreTest {
             assertTrue(observedState.contains(expectedProgressState3))
         }
 
+    @Test
+    fun `GIVEN analysis succeeds WHEN deletedProduct is true and deletedProductReported is false THEN product not available is shown`() =
```
Comment (rahulsainani) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1286]: GIVEN deletedProduct is true and deletedProductReported is false WHEN report product back in stock is clicked THEN thanks for reporting card is shown

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt [hunk 12]
```diff
@@ -1282,6 +1282,55 @@ class ReviewQualityCheckStoreTest {
...
+                                deletedProductReported = false,
+                            )
+                        },
+                        reanalysis = AnalysisStatusDto.PENDING,
+                        statusProgress = {
+                            AnalysisStatusProgressDto(
+                                status = AnalysisStatusDto.COMPLETED,
+                                progress = 100.0,
+                            )
+                        },
```
Comment (rahulsainani) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1304]: 🔍 These two might not need to be set for this test, seem unrelated

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt (hunk 2 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt]: 👌

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt (hunk 11 above)
Comment (rahulsainani) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1286]:
```suggestion
    fun `GIVEN deletedProduct is true and deletedProductReported is false WHEN report product back in stock is clicked THEN thanks for reporting card is shown`() =
```

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt [hunk 13]
```diff
@@ -1282,6 +1282,55 @@ class ReviewQualityCheckStoreTest {
...
+                    ),
+                    networkChecker = FakeNetworkChecker(isConnected = true),
```
Comment (rahulsainani) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1306]: Let's set report to report created so we know this only should happen when report created is returned. On returning another response, UI should react differently. OOC, what do we show when not deleted is returned?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt (hunk 9 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt]: Removed default value `it.asEnumOrDefault<ReportBackInStockStatusDto>()`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt (hunk 10 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt]: lol updated doc to `report created`

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt (hunk 12 above)
Comment (calumozilla) [reply fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt]: Ya this isn't necessary, updated to `{ null }`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt (hunk 7 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt]: I moved the `when` logic back into the above function, but didn't add the data class, since we would still have to check if it's `Error.ProductNotAvailable` to create a ReviewQualityCheckInfoCard with button. It only seemed marginally better. But if you have something else in mind, maybe I can add as a follow up, if we're trying to get this in before code freeze.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 8 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]: hm ya we do update to `ThanksForReporting` immediately, I removed it here

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt (hunk 13 above)
Comment (calumozilla) [reply fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1303]: Updated to return REPORT_CREATED for this test, also added a test to return NOT_DELETED which will return no analysis state since that's what desktop is doing

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt (hunk 7 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ProductAnalysisError.kt]: That's fair, we can discuss that approach after landing this 👍

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 8 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]: I'm not sure we should not update the state to `NoAnalysisPresent()` directly here. By just doing this, we are expecting that the report product api responses are immediately in sync with the analysis api.  This case should hardly ever happen. 

I think for the `productReviewState` we should use the analysis api as the source of truth  i.e fire the Fetch action that will take care of fetching and mapping, so the mapper does the usual checks and maps to the correct state - which could be `NoAnalysisPresent`. wdyt?

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt (hunk 13 above)
Comment (rahulsainani) [reply fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1303]: Thanks! 

For not deleted, see [this](https://github.com/mozilla-mobile/firefox-android/pull/4756#discussion_r1425158603).

**Review by rahulsainani — APPROVED** [review]
This is looking good and works really well! 🎉 

Please check this [comment](https://github.com/mozilla-mobile/firefox-android/pull/4756#discussion_r1425158603). It's the only thing I'm unsure of right now. It's an implementation "idea" that shouldn't change the behavior and since It's a rare edge case, we can even do this as a follow up.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 8 above)
Comment (calumozilla) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]: Hm okay, but won't fetching and calling the mapper still map to one of the errors Error.ProductAlreadyReported or ProductReviewState.Error.ProductNotAvailable?

Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:90]:
```
if (status == ReportBackInStockStatusDto.NOT_DELETED) {
    // fetch and map
}
```

Since this will happen when the report api returns not deleted, I doubt that the analysis api will return a response that leads to Error.ProductAlreadyReported or ProductReviewState.Error.ProductNotAvailable. 

If that does indeed happen, it would also happen if we close and open the sheet as that would trigger the fetch and map again.

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt [hunk 14]
```diff
@@ -1283,6 +1284,115 @@ class ReviewQualityCheckStoreTest {
...
+
+            val expected = ReviewQualityCheckState.OptedIn(
+                productReviewState = ReviewQualityCheckState.OptedIn.ProductReviewState.Error.ProductNotAvailable,
+                productRecommendationsPreference = false,
+                productRecommendationsExposure = true,
+                productVendor = ProductVendor.BEST_BUY,
+            )
+            assertEquals(expected, tested.state)
+
+            val noAnalysisTested = ReviewQualityCheckStore(
```
Comment (rahulsainani) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1367]: As a future improvement, we can use 
```
productAnalysis = { count -> 
// initial analysis with deleted product
// subsequent analysis with no analysis 
 }  
```
as fake response instead of creating two different stores, so we're testing on the same instance of the store. 

Alternatively you can always set an initialState and then just test the logic from here on.

Comment (calumozilla) [reply fenix/app/src/test/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStoreTest.kt:1367]: Hey, coming back to this, how would the state count update?
